File: test/java/brs/assetexchange/AssetAccountServiceImplTest.java
package brs.assetexchange;

import brs.Account.AccountAsset;
import brs.Asset;
import brs.db.store.AccountStore;
import org.junit.Before;
import org.junit.Test;

import java.util.ArrayList;

import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class AssetAccountServiceImplTest {

  private AssetAccountServiceImpl t;

  private AccountStore mockAccountStore;

  @Before
  public void setUp() {
    mockAccountStore = mock(AccountStore.class);

    t = new AssetAccountServiceImpl(mockAccountStore);
  }

  @Test
  public void getAssetAccounts() {
    final int from = 1;
    final int to = 5;
    final Asset mockAsset = mock(Asset.class);
    final ArrayList<AccountAsset> mockAccountIterator = new ArrayList<>();
    
    when(mockAccountStore.getAssetAccounts(eq(mockAsset), eq(false), eq(0L), eq(from), eq(to))).thenReturn(mockAccountIterator);

    assertEquals(mockAccountIterator, t.getAssetAccounts(mockAsset, false, 0L, from, to));
  }

  @Test
  public void getAssetAccountsCount() {
    when(mockAccountStore.getAssetAccountsCount(eq(123L), eq(0L))).thenReturn(5);

    assertEquals(5L, t.getAssetAccountsCount(123, eq(0L)));
  }
}

File: src/test/kotlin/brs/assetexchange/AssetAccountServiceImplTest.kt
package brs.assetexchange

import brs.entity.Account.AccountAsset
import brs.db.AccountStore
import brs.services.impl.AssetAccountServiceImpl
import io.mockk.mockk
import io.mockk.every
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test

class AssetAccountServiceImplTest {

    private lateinit var t: AssetAccountServiceImpl

    private lateinit var mockAccountStore: AccountStore

    @Before
    fun setUp() {
        mockAccountStore = mockk()
        t = AssetAccountServiceImpl(mockAccountStore)
    }

    @Test
    fun getAssetAccounts() {
        val assetId = 4L
        val from = 1
        val to = 5

        val mockAccountIterator = mockk<Collection<AccountAsset>>()

        every { mockAccountStore.getAssetAccounts(eq(assetId), eq(from), eq(to)) } returns mockAccountIterator

        assertEquals(mockAccountIterator, t.getAssetAccounts(assetId, from, to))
    }

    @Test
    fun getAssetAccounts_withHeight() {
        val assetId = 4L
        val from = 1
        val to = 5
        val height = 3

        val mockAccountIterator = mockk<Collection<AccountAsset>>()

        every { mockAccountStore.getAssetAccounts(eq(assetId), eq(height), eq(from), eq(to)) } returns mockAccountIterator

        assertEquals(mockAccountIterator, t.getAssetAccounts(assetId, height, from, to))
    }

    @Test
    fun getAssetAccounts_withHeight_negativeHeightGivesForZeroHeight() {
        val assetId = 4L
        val from = 1
        val to = 5
        val height = -2

        val mockAccountIterator = mockk<Collection<AccountAsset>>()

        every { mockAccountStore.getAssetAccounts(eq(assetId), eq(from), eq(to)) } returns mockAccountIterator

        assertEquals(mockAccountIterator, t.getAssetAccounts(assetId, height, from, to))
    }

    @Test
    fun getAssetAccountsCount() {
        every { mockAccountStore.getAssetAccountsCount(eq(123L)) } returns 5

        assertEquals(5L, t.getAssetAccountsCount(123).toLong())
    }
}
